tests/integration_test_net.go

```diff
@@ -169,6 +177,66 @@ type integrationTestNode struct {
 	done      <-chan struct{}
 }
 
+////////////////////////////////////////////////////////////////////////////////
+// Memory profiler.
+// if enabled with the `-test.heap.profile` flag, it will write a heap dump to
+// the `../build/profile/` directory at the end of the test run.
+// The file will be named `mem_<test_name>.pprof` where `<test_name>` is the name
+// of the test that started the profiling.
+////////////////////////////////////////////////////////////////////////////////
+
+type netProfiler struct {
+	tb testing.TB
+	// highest memory usage seen so far,
+	// used to write only the peak consumption to a file
+	highestSeen uint64
+}
+
+// since `go test` already parses the flag, this variable needs to be declared
+// at the package level, so that it can be parsed when go test parses other flags.
+var testProfileEnabled = flag.Bool("test.heap.profile", false,
+	`Enable memory profiling for integration tests. 
+	If set, a memory profile will be written to the build/profile directory 
+	at the end of the test run.`)
+
+func profileRoutine(ctx context.Context, n *netProfiler) {
+
+	n.highestSeen = 0
+
+	buffer := bytes.NewBuffer(nil)
+	memStats := &runtime.MemStats{}
+
+	ticker := time.NewTicker(100 * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			runtime.ReadMemStats(memStats)
+			if memStats.HeapAlloc <= n.highestSeen {
+				continue
+			}
+			buffer.Reset()
+			n.highestSeen = memStats.HeapAlloc
+			require.NoError(n.tb, pprof.WriteHeapProfile(buffer))
+
+		case <-ctx.Done():
+			// write a file with the name of the test case that started the profiling
+			buildProfile := "../build/profile/"
```

Comment:
I am not a particular fan of this relative addressing. It can mess up things if you run tests from a different working directory. It should either be an absolute directory, relative to the test-case file, or in the current working directory.